### DISCUSSION TITLE
How to retrieve a list of all existing Worfkflow Instances

### DISCUSSION THREAD

**Discussion by ckolumbus** [danielgerlag/workflow-core#861] [Q&A]
Hi *
I really like this workflow engine: small  & intuitive to use. There's one issue I stumbled over : In `IWorkflowRepository` there exists an API to retrieve all instances at once by setting all query parameters to `null` or `""`  [`IWorkflowRepository.cs`](https://github.com/danielgerlag/workflow-core/blob/06470ce83446a970ff95f2f070793cefa7f6e5cc/src/WorkflowCore/Interface/Persistence/IWorkflowRepository.cs#L18)
```csharp
Task<IEnumerable<WorkflowInstance>> GetWorkflowInstances(WorkflowStatus? status, string type, DateTime? createdFrom, DateTime? createdTo, int skip, int take);
```
BUT this API is marked `obsolete`. 

The new existing API takes a list if `id`s, which seem to be the GUIDs of the running workflows:

```csharp
 Task<IEnumerable<WorkflowInstance>> GetWorkflowInstances(IEnumerable<string> ids, CancellationToken cancellationToken = default);
```

 As this is exactly the information I'm looking for (IDs of all running instances) I'm stuck now on where to get the information formerly provided by the obsolete API.

Any hints on this?

**Comment by wallee85** [comment]
you may need a elasticsearch extensions to do this,i think

**Reply by ckolumbus** [reply]
@pangjianxin thanks for the answer. IMHO this would be unfortunate. I can't use elastic search due to project requirements, and frankly, I don't want to either just for getting a list of workflows.

**Reply by neozhu** [reply]
Especially in the manual workflow needs to query my pending workflow and I have completed the workflow

**Comment by bainewedlock** [comment]
Guess I'm somewhat late to the party.
But I just recently found this lib and it might improve a lot of our projects. (Alot better than any other projects that I checked out so far.)

**Would it make sense to create new pull requests?**

Idea 1: Remove the [Obsolete] attribute from `GetWorkflowInstances`. Not sure why it is marked. Guess the parameters could be better, but the function is still useful as it is.

Idea 2: Add a function to purge old workflow instances. (complete or terminated workflows which are older than X days)

**Comment by fradaham** [comment]
I'm checking out this library, and like it. But exactly this feature, getting all workflow instances of a certain type and status,  is something I need (seems like a basic thing), and I don't like to let my application rest upon an obsolete method in a library dependency. Why is it obsolete? I why should you need elasticsearch to retrieve this information? It should be readily retrievable from the persistence provider. 

I actually tried to add the elasticsearch dependency, but couldn't get it working. I don't want an unnecessary dependency to start with, and point was proven when I couldn't get it to function.
